pandas/core/groupby/generic.py

```diff
@@ -1579,27 +1658,66 @@ def _insert_inaxis_grouper_inplace(self, result):
             if in_axis:
                 result.insert(0, name, lev)
 
-    def _wrap_aggregated_output(self, output, names=None):
-        agg_axis = 0 if self.axis == 1 else 1
-        agg_labels = self._obj_with_exclusions._get_axis(agg_axis)
+    def _wrap_aggregated_output(
+        self, output: Mapping[int, Union[Series, np.ndarray]], columns: Index
+    ) -> DataFrame:
+        """
+        Wraps the output of DataFrameGroupBy aggregations into the expected result.
+
+        Parameters
+        ----------
+        output : dict[int, Union[Series, np.ndarray]]
+            Dict where the key represents the columnar-index and the values are
+            the actual results.
+        columns : pd.Index
+            Column names to apply
+
+        Returns
+        -------
+        DataFrame
+        """
+        assert list(output.keys()) == sorted(output.keys())
 
-        output_keys = self._decide_output_index(output, agg_labels)
+        result = DataFrame(output)
+        result.columns = columns
 
         if not self.as_index:
-            result = DataFrame(output, columns=output_keys)
             self._insert_inaxis_grouper_inplace(result)
             result = result._consolidate()
         else:
             index = self.grouper.result_index
-            result = DataFrame(output, index=index, columns=output_keys)
+            result.index = index
 
         if self.axis == 1:
             result = result.T
 
         return self._reindex_output(result)._convert(datetime=True)
 
-    def _wrap_transformed_output(self, output, names=None) -> DataFrame:
-        return DataFrame(output, index=self.obj.index)
+    def _wrap_transformed_output(
+        self, output: Mapping[int, Union[Series, np.ndarray]], columns: Index
```

Comment:
here and elsewhere the `Union[Series, np.ndarray]` never gets EA?

Reply:
I _think_ it does.

I had a go at trying to avoid the cast, and needed an instance EA in _wrap_series_output to get the tests to pass.

```diff
diff --git a/pandas/core/groupby/generic.py b/pandas/core/groupby/generic.py
index 92378ca91..20cbc9c96 100644
--- a/pandas/core/groupby/generic.py
+++ b/pandas/core/groupby/generic.py
@@ -20,6 +20,7 @@ from typing import (
     Type,
     Union,
     cast,
+    overload,
 )
 import warnings
 
@@ -50,6 +51,7 @@ from pandas.core.dtypes.missing import _isna_ndarraylike, isna, notna
 
 from pandas._typing import FrameOrSeries
 import pandas.core.algorithms as algorithms
+from pandas.core.arrays import ExtensionArray
 from pandas.core.base import DataError, SpecificationError
 import pandas.core.common as com
 from pandas.core.frame import DataFrame
@@ -332,9 +334,29 @@ class SeriesGroupBy(GroupBy):
 
         return DataFrame(results, columns=columns)
 
+    # TODO: _OUTPUT should be Union[Series, np.ndarray, ExtensionArray] but since
+    # np.ndarray resolves to Any, can't be used with overloads without producing
+    # error: Overloaded function signatures 1 and 2 overlap with incompatible return
+    #  types
+    # while np.ndarray resolves to Any, it will be compatible with Series and
+    #  ExtensionArray
+    _OUTPUT = Union[Series, ExtensionArray]
+
+    @overload
+    def _wrap_series_output(
+        self, output: Mapping[int, _OUTPUT], index: Index, columns: Index
+    ) -> DataFrame:
+        ...
+
+    @overload  # noqa: F811
     def _wrap_series_output(
+        self, output: _OUTPUT, index: Index, columns: Index
+    ) -> Series:
+        ...
+
+    def _wrap_series_output(  # noqa: F811
         self,
-        output: Mapping[int, Union[Series, np.ndarray]],
+        output: Union[_OUTPUT, Mapping[int, _OUTPUT]],
         index: Index,
         columns: Index,
     ) -> Union[Series, DataFrame]:
@@ -360,17 +382,18 @@ class SeriesGroupBy(GroupBy):
         In the vast majority of cases output and columns will only contain one
         element. The exception is operations that expand dimensions, like ohlc.
         """
-        assert len(output) == len(columns)
-        assert list(output.keys()) == sorted(output.keys())
 
-        result: Union[Series, DataFrame]
-        if len(output) > 1:
+        if (
+            not isinstance(output, (Series, np.ndarray, ExtensionArray))
+            and len(output) > 1
+        ):
+            assert len(output) == len(columns)
+            assert list(output.keys()) == sorted(output.keys())
             result = DataFrame(output, index=index)
             result.columns = columns
+            return result
         else:
-            result = Series(output[0], index=index, name=columns[0])
-
-        return result
+            return Series(output, index=index, name=columns[0])
 
     def _wrap_aggregated_output(
         self, output: Mapping[int, Union[Series, np.ndarray]], columns: Index
@@ -397,8 +420,13 @@ class SeriesGroupBy(GroupBy):
         """
         assert list(output.keys()) == sorted(output.keys())
 
+        if len(output) == 1:
+            _output = output[0]
+        else:
+            _output = output
+
         result = self._wrap_series_output(
-            output=output, index=self.grouper.result_index, columns=columns
+            output=_output, index=self.grouper.result_index, columns=columns
         )
         return self._reindex_output(result)._convert(datetime=True)
 
@@ -426,13 +454,12 @@ class SeriesGroupBy(GroupBy):
         """
         assert list(output.keys()) == sorted(output.keys())
 
+        assert len(output) == 1
+
         result = self._wrap_series_output(
-            output=output, index=self.obj.index, columns=columns
+            output=output[0], index=self.obj.index, columns=columns
         )
 
-        # No transformations increase the ndim of the result
-        # Unfortunately need to cast for mypy to know this
-        result = cast(Series, result)
         return result
 
     def _wrap_applied_output(self, keys, values, not_indexed_same=False):
```

Reply:
p.s. I don't recommend the above since still needed `len(output) > 1` in the condition so the return type could not be established with the argument type alone and therefore overloading is not the answer here.